src/Data/Array/Polarized/Pull/Internal.hs

```diff
@@ -0,0 +1,73 @@
+{-# OPTIONS_GHC -fno-warn-partial-type-signatures #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+
+module Data.Array.Polarized.Pull.Internal where
+
+import qualified Data.Functor.Linear as Data
+import Prelude.Linear
+import qualified Prelude
+import Data.Vector (Vector)
+import qualified Data.Vector as Vector
+
+import qualified Unsafe.Linear as Unsafe
+
+-- | A pull array represents arrays which are easy to extract elements from.
+-- They provide an indexing function, and a length. If the indexing function
+-- is called at a point out of range, the result is unspecified.
+data Array a where
+  Array :: (Int -> a) -> Int -> Array a
+-- Int is movable, so this is the same as Array :: (Int -> a) -> ...
+
+instance Data.Functor Array where
+  fmap f (Array g n) = Array (\x -> f (g x)) n
+
+-- XXX: This doesn't match how PullArrays should work, but it is accepted
```

Comment:
The type of pull arrays is abstract specifically because some things which are accepted are not valid. So I'm not sure it's a problem.

Reply:
This itself isn't a problem, but it's concerning that `singleton` below isn't accepted...